### PR TITLE
1727 refactor listener for market participant int. events

### PR DESCRIPTION
**Pull request by jonasdmoeller**
## Description

Refactored integrating actor created events, and added support for actor status updated and external id changed.

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/IMarketParticipantUpdatedCommandHandler.cs [hunk 1]
```diff
@@ -13,12 +13,18 @@
...
 
 using System.Threading.Tasks;
-using GreenEnergyHub.Charges.Domain.Dtos.MarketParticipantsUpdatedEvents;
+using GreenEnergyHub.Charges.Domain.Dtos.Events;
 
 namespace GreenEnergyHub.Charges.Application.MarketParticipants.Handlers
 {
-    public interface IMarketParticipantCreatedHandler
+    /// <summary>
+    /// Handle updated for market participants
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/IMarketParticipantUpdatedCommandHandler.cs:21]:
```suggestion
    /// Handle updated events for market participants
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/Mappers/ActorIntegrationEventMapper.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+using Energinet.DataHub.MarketParticipant.Integration.Model.Parsers;
+using GreenEnergyHub.Charges.Domain.Dtos.Events;
+
+namespace GreenEnergyHub.Charges.Application.MarketParticipants.Handlers.Mappers
+{
+    public class ActorIntegrationEventMapper : IActorIntegrationEventMapper
+    {
+        private readonly ISharedIntegrationEventParser _sharedIntegrationEventParser;
+
+        public ActorIntegrationEventMapper(ISharedIntegrationEventParser sharedIntegrationEventParser)
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/Mappers/ActorIntegrationEventMapper.cs:26]:
```suggestion
        public MarketParticipantIntegrationEventMapper(ISharedIntegrationEventParser sharedIntegrationEventParser)
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/Mappers/ActorIntegrationEventMapper.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Linq;
+using Energinet.DataHub.MarketParticipant.Integration.Model.Dtos;
+using Energinet.DataHub.MarketParticipant.Integration.Model.Parsers;
+using GreenEnergyHub.Charges.Domain.Dtos.Events;
+
+namespace GreenEnergyHub.Charges.Application.MarketParticipants.Handlers.Mappers
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/Mappers/ActorIntegrationEventMapper.cs:20]: This file do not have good test coverage

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/Mappers/IActorIntegrationEventMapper.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using GreenEnergyHub.Charges.Domain.Dtos.Events;
+
+namespace GreenEnergyHub.Charges.Application.MarketParticipants.Handlers.Mappers
+{
+    public interface IActorIntegrationEventMapper
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/Mappers/IActorIntegrationEventMapper.cs:19]:
```suggestion
    public interface IMarketParticipantIntegrationEventMapper
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/Mappers/IActorIntegrationEventMapper.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        MarketParticipantCreatedCommand MapFromActorCreated(byte[] message);
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/Mappers/IActorIntegrationEventMapper.cs:21]:
```suggestion
        MarketParticipantCreatedCommand MapFromMarketParticipantCreated(byte[] message);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantB2CActorIdCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+// limitations under the License.
+
+using System;
+using System.Threading.Tasks;
+using GreenEnergyHub.Charges.Domain.Dtos.Events;
+using GreenEnergyHub.Charges.Domain.MarketParticipants;
+
+namespace GreenEnergyHub.Charges.Application.MarketParticipants.Handlers
+{
+    public class MarketParticipantB2CActorIdChangedCommandHandler : IMarketParticipantB2CActorIdChangedCommandHandler
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantB2CActorIdCommandHandler.cs:22]: Cant see unit test of this class

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/IMarketParticipantCreatedCommandHandler.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+{
+    /// <summary>
+    /// Handle command for creating new market participants
+    /// </summary>
+    public interface IMarketParticipantCreatedCommandHandler
+    {
+        /// <summary>
+        /// Handle MarketParticipantCreatedCommand
+        /// </summary>
+        Task HandleAsync(MarketParticipantCreatedCommand marketParticipantUpdatedCommand);
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/IMarketParticipantCreatedCommandHandler.cs:28]:
```suggestion
        Task HandleAsync(MarketParticipantCreatedCommand marketParticipantCreatedCommand);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantStatusChangedCommandHandler.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+// limitations under the License.
+
+using System;
+using System.Threading.Tasks;
+using GreenEnergyHub.Charges.Domain.Dtos.Events;
+using GreenEnergyHub.Charges.Domain.MarketParticipants;
+
+namespace GreenEnergyHub.Charges.Application.MarketParticipants.Handlers
+{
+    public class MarketParticipantStatusChangedCommandHandler : IMarketParticipantStatusChangedCommandHandler
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/MarketParticipantStatusChangedCommandHandler.cs:22]: No unittests

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/IMarketParticipantUpdatedCommandHandler.cs [hunk 9]
```diff
@@ -13,12 +13,18 @@
...
+    /// </summary>
+    public interface IMarketParticipantUpdatedCommandHandler
     {
-        Task HandleAsync(MarketParticipantUpdatedEvent marketParticipantUpdatedEvent);
+        /// <summary>
+        /// Handle market participant update command
+        /// </summary>
+        Task HandleAsync(MarketParticipantUpdatedCommand message);
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/MarketParticipants/Handlers/IMarketParticipantUpdatedCommandHandler.cs:28]:
```suggestion
        Task HandleAsync(MarketParticipantUpdatedCommand command);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Common/EnvironmentSettingNames.cs [hunk 10]
```diff
@@ -66,6 +66,8 @@ public static class EnvironmentSettingNames
         // Integration, marketparticipant/actor domain
         public const string MarketParticipantChangedSubscriptionName = "MARKET_PARTICIPANT_CHANGED_SUBSCRIPTION_NAME";
         public const string MarketParticipantCreatedSubscriptionName = "MARKET_PARTICIPANT_CREATED_SUBSCRIPTION_NAME";
+        public const string MarketParticipantStatusChangedSubscriptionName = "MARKET_PARTICIPANT_STATUS_CHANGED_SUBSCRIPTION_NAME";
+        public const string MarketParticipantExternalActorIdChangedSubscriptionName = "MARKET_PARTICIPANT_STATUS_EXTERNAL_ACTOR_ID_CHANGED_SUBSCRIPTION_NAME";
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Common/EnvironmentSettingNames.cs:70]: No subscription is created, and this variable should be added to local.settings sample file

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantB2CActorIdEndpoint.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        [Function(FunctionName)]
+        public async Task RunAsync([ServiceBusTrigger(
+            "%" + EnvironmentSettingNames.IntegrationEventTopicName + "%",
+            "%" + EnvironmentSettingNames.MarketParticipantExternalActorIdChangedSubscriptionName + "%",
+            Connection = EnvironmentSettingNames.DataHubListenerConnectionString)]
+            byte[] message)
+        {
+            var externalIdChanged = _actorIntegrationEventMapper.MapFromActorExternalIdChanged(message);
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantB2CActorIdEndpoint.cs:48]:
```suggestion
            var externalIdChangedCommand = _actorIntegrationEventMapper.MapFromActorExternalIdChanged(message);
```
Hard to guess what this is from the current naming, so either this or use full name e.g. marketParticipantB2CActorIdChangedCommand

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantStatusChangedEndpoint.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        [Function(FunctionName)]
+        public async Task RunAsync([ServiceBusTrigger(
+            "%" + EnvironmentSettingNames.IntegrationEventTopicName + "%",
+            "%" + EnvironmentSettingNames.MarketParticipantStatusChangedSubscriptionName + "%",
+            Connection = EnvironmentSettingNames.DataHubListenerConnectionString)]
+            byte[] message)
+        {
+            var updateStatusCommand = _actorIntegrationEventMapper.MapFromActorStatusChanged(message);
+            await _marketParticipantStatusChangedCommandHandler.HandleAsync(updateStatusCommand).ConfigureAwait(false);
+            await _unitOfWork.SaveChangesAsync().ConfigureAwait(false);
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantStatusChangedEndpoint.cs:50]: Should this logic be in an endpoint? And if it should it should be unittested, I would put the unitofwork into the handler, an endpoint should only map items and parse on those to handlers

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/MarketParticipants/Handlers/MarketParticipantDomainEventMapperTests.cs [hunk 13]
```diff
@@ -18,6 +18,7 @@
 using Energinet.DataHub.MarketParticipant.Integration.Model.Dtos;
 using FluentAssertions;
 using GreenEnergyHub.Charges.Application.MarketParticipants.Handlers;
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/MarketParticipants/Handlers/MarketParticipantDomainEventMapperTests.cs:20]: not needed

Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/MarketParticipants/Handlers/MarketParticipantDomainEventMapperTests.cs:20]: Filename should match the file its testing which I suggested should be renamed, so this should be MarketParticipantIntegrationEventMapperTests

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/MarketParticipants/Handlers/MarketParticipantEventHandlerTests.cs [hunk 14]
```diff
@@ -16,10 +16,9 @@
 using System.Collections.Generic;
 using System.Threading.Tasks;
 using AutoFixture.Xunit2;
-using Energinet.DataHub.MarketParticipant.Integration.Model.Dtos;
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/MarketParticipants/Handlers/MarketParticipantEventHandlerTests.cs:19]: Classname should match file its testing

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/MarketParticipants/Handlers/MarketParticipantRoleMapperTests.cs [hunk 15]
```diff
@@ -13,9 +13,12 @@
 // limitations under the License.
 
 using System;
+using System.Collections.Generic;
+using System.Linq;
 using Energinet.DataHub.MarketParticipant.Integration.Model.Dtos;
 using FluentAssertions;
 using GreenEnergyHub.Charges.Application.MarketParticipants.Handlers;
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/MarketParticipants/Handlers/MarketParticipantRoleMapperTests.cs:20]:
```suggestion
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantStatusChangedEndpoint.cs (hunk 12 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/MarketParticipant/MarketParticipantStatusChangedEndpoint.cs:53]: The unit of work save should be as close as possible to the entry point of the application.

**Review by HenrikSommer — APPROVED** [review]
Approved based on @Mech0z's comments and the good job with the code.

**Pull request merged**